### ISSUE TITLE
Be open to extension

### ISSUE THREAD

**Issue by tomkerkhove** [arcus-azure/arcus#143]
We provide a lot of great stuff, but when you want to extend or tweak something to your needs it becomes harder because we are not always open to extension.

We should ensure that we allow people to derive and extend things by using `protected virtual` instead of `private`, for example.

**Comment by stijnmoreels** [comment]
I agree, though there may be places were it's recommended to privatize stuff. Maybe this is a bit subjective. Useful places that could be extended, For sure!

**Comment by fgheysels** [comment]
I agree with @stijnmoreels : you don't want to change all private's to protected virtual by default.   Same is true for public members: you don't want to change them all to virtual by default.
Moreover, I used to never specify public members as `virtual`.   If I want to allow a consumer to change the behavior of a method, I mostly design it like this:

```csharp
public void DoSomething( Foo bar )
{
    if( bar == null )
    {
        throw new ArgumentNullException(nameof(bar));
    }

   DoSomethingCore(bar);
}

protected virtual void DoSomething(Foo bar)
{
    ... 
}
```

see also  [this](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/virtual-members)  article for design guidelines on using `virtual`.

Do we have specific cases where we're lacking extensibility ?

(ps: should this be a discussion ? )

**Comment by tomkerkhove** [comment]
It's not a discussion because it's a must :D Today this is a pain.

I agree that we shouldn't do it for all privates, that was not really my intend indeed. However,  I would personally keep it simple and make the protected methods virtual itself rather than different sub pieces since you can still manage that through base.MethodName, no?

Otherwise it becomes too complex imo.

**Comment by fgheysels** [comment]
My intention was rather: should it be a 'discussion' so that we can agree on how to do it.  :)

When looking at the [`KeyVaultCachedSecretProvider`](https://github.com/arcus-azure/arcus.security/blob/master/src/Arcus.Security.Providers.AzureKeyVault/KeyVaultCachedSecretProvider.cs) class, you'll see that this class has one method which is `public virtual`: `StoreSecretAsync`.   In this specific case, I actually wonder why this method is virtual.  As we're in the `KeyVaultCachedSecretProvider`, you'll certainly want to store the secret in KeyVault, so why would anyone really want to override the functionality ?  
If you want to tweak something, it'll probably be some additional logic that must be executed before you store the secret, or after you've stored the secret.   Hence my suggestion to have something like this:

```csharp
public async Task StoreSecretAsync( ... )
{
    await OnBeforeStoreSecretAsync(...);

    await _secretProvider.StoreSecretAsync( ... );
   MemoryCache.Set(... );

    await OnAfterStoreSecretAsync(...);
}

protected virtual Task OnBeforeStoreSecretAsync() {}
```

Instead of providing protected virtual methods as extension points, you could also implement this via events.  However, that might be a bit of a pain when working with async code.

Anyhow, I think it's to blunt to just say that all protected methods must be virtual.   Take a look at the [`KeyVaultSecretProvider`](https://github.com/arcus-azure/arcus.security/blob/master/src/Arcus.Security.Providers.AzureKeyVault/KeyVaultSecretProvider.cs) class.
One `protected` method: `GetSecretClient` which -imho- doesn't make sense to override.
Two `static protected` methods:  You can't override static, and IMHO it also doesn't make sense to override them:
- `ThrottleTooManyRequestsAsync`   : You can allow users to change the throttling behavior by havving a settable `ThrottlingPolicy` property f.i.
- (The other protected static is the same method with a different signature, so that would mean that if you override the first one, you must also override the 2nd one.  Having an adaptable 'policy' which the consumer can specify, makes more sense imo.

**Comment by stijnmoreels** [comment]
The `KeyVaultCachedSecretProvider` is maybe not a good example to use here. The reason btw the store method is virtual, is because when using KeyVault with caching, there's some extra methods with the boolean `ignoreCache` which are only available in the cached variantes.
But again, I think not a good example to use here.

The `KeyVaultSecretProvider` itself or even this open PR #282 may be better ones. I think the goal here is to make sure that virtualize public members (coming from interface or otherwise) and provide access (by making it `protected`) to otherwise private members. In the #282, you'll see the consumer having access to the user-configurable information and the actual secret-retrieval functionality. Making it open for extension (`public virtual` from interface) but closed to modification (`private` members with the hard-lifting becomes `protected` but not `virtual`).

**Comment by stijnmoreels** [comment]
As stated by @fgheysels here https://github.com/arcus-azure/arcus.security/pull/282#discussion_r671077238 : we may not be ready to know what the consumers want. On the other hand, if we continue the approach we're using, they are free to use everything and set their own limitations instead of us trying to restrict something we can't predict.
